File: lib_arch/src/main/java/com/yyxnb/arch/delegate/ActivityLifecycle.java
package com.yyxnb.arch.delegate;

import android.app.Activity;
import android.app.Application;
import android.os.Bundle;
import android.support.v4.app.FragmentActivity;
import android.util.LruCache;

import com.yyxnb.arch.base.IActivity;
import com.yyxnb.arch.utils.AppManager;

public class ActivityLifecycle implements Application.ActivityLifecycleCallbacks {

    private volatile static ActivityLifecycle lifecycle = null;

    public static ActivityLifecycle getInstance() {
        if (lifecycle == null) {
            synchronized (ActivityLifecycle.class) {
                if (lifecycle == null) {
                    lifecycle = new ActivityLifecycle();
                }
            }
        }
        return lifecycle;
    }

    private LruCache<String, IActivityDelegate> cache = new LruCache<>(100);

    @Override
    public void onActivityCreated(Activity activity, Bundle savedInstanceState) {
        if (activity != null) {
            AppManager.getInstance().addActivity(activity);
        }

        if (activity instanceof IActivity) {

            IActivityDelegate activityDelegate = null;

            if (fetchActivityDelegate(activity) != null) {
                activityDelegate = fetchActivityDelegate(activity);
            } else {
                activityDelegate = newDelegate(activity);
            }

            cache.put(getKey(activity), activityDelegate);

            activityDelegate.onCreate(savedInstanceState);

            registerFragmentCallback(activity);
        }

    }

    private void registerFragmentCallback(Activity activity) {

        if (activity instanceof FragmentActivity) {
            ((FragmentActivity) activity).getSupportFragmentManager().registerFragmentLifecycleCallbacks(FragmentLifecycle.getInstance(), true);
        }

    }

    @Override
    public void onActivityStarted(Activity activity) {
        if (fetchActivityDelegate(activity) != null) {
            fetchActivityDelegate(activity).onStart();
        }
    }

    @Override
    public void onActivityResumed(Activity activity) {
        if (fetchActivityDelegate(activity) != null) {
            fetchActivityDelegate(activity).onResume();
        }
    }

    @Override
    public void onActivityPaused(Activity activity) {
        if (fetchActivityDelegate(activity) != null) {
            fetchActivityDelegate(activity).onPause();
        }
    }

    @Override
    public void onActivityStopped(Activity activity) {
        if (fetchActivityDelegate(activity) != null) {
            fetchActivityDelegate(activity).onStop();
        }
    }

    @Override
    public void onActivitySaveInstanceState(Activity activity, Bundle outState) {
        if (fetchActivityDelegate(activity) != null) {
            fetchActivityDelegate(activity).onSaveInstanceState(activity, outState);
        }
    }

    @Override
    public void onActivityDestroyed(Activity activity) {
        if (activity != null) {
            AppManager.getInstance().removeActivity(activity);
        }
        if (fetchActivityDelegate(activity) != null) {
            fetchActivityDelegate(activity).onDestroy();
        }
    }

    private IActivityDelegate fetchActivityDelegate(Activity activity) {

        if (activity instanceof IActivity) {
            return cache.get(getKey(activity));
        }
        return null;
    }

    private IActivityDelegate newDelegate(Activity activity) {

        return new ActivityDelegateImpl((FragmentActivity) activity);
    }

    private String getKey(Activity activity) {
        return activity.getClass().getName() + activity.hashCode();
    }
}

File: lib_arch/src/main/java/com/yyxnb/arch/delegate/ActivityLifecycle.kt
package com.yyxnb.arch.delegate

import android.app.Activity
import android.app.Application.ActivityLifecycleCallbacks
import android.os.Bundle
import android.util.LruCache
import androidx.fragment.app.FragmentActivity
import com.yyxnb.arch.base.IActivity
import com.yyxnb.arch.utils.AppManager.addActivity
import com.yyxnb.arch.utils.AppManager.removeActivity

/**
 * Activity 注册监听生命周期
 *
 * @author yyx
 */
object ActivityLifecycle : ActivityLifecycleCallbacks {

    private val cache = LruCache<String, IActivityDelegate?>(100)

    override fun onActivityCreated(activity: Activity, savedInstanceState: Bundle?) {
        activity?.let { addActivity(it) }
        if (activity is IActivity) {
            var activityDelegate: IActivityDelegate? = null
            activityDelegate = if (fetchActivityDelegate(activity) != null) {
                fetchActivityDelegate(activity)
            } else {
                newDelegate(activity)
            }
            cache.put(getKey(activity), activityDelegate)
            activityDelegate!!.onCreate(savedInstanceState)
            registerFragmentCallback(activity)
        }
    }

    private fun registerFragmentCallback(activity: Activity) {
        if (activity is FragmentActivity) {
            activity.supportFragmentManager.registerFragmentLifecycleCallbacks(FragmentLifecycle, true)
        }
    }

    override fun onActivityStarted(activity: Activity) {
        if (fetchActivityDelegate(activity) != null) {
            fetchActivityDelegate(activity)!!.onStart()
        }
    }

    override fun onActivityResumed(activity: Activity) {
        if (fetchActivityDelegate(activity) != null) {
            fetchActivityDelegate(activity)!!.onResume()
        }
    }

    override fun onActivityPaused(activity: Activity) {
        if (fetchActivityDelegate(activity) != null) {
            fetchActivityDelegate(activity)!!.onPause()
        }
    }

    override fun onActivityStopped(activity: Activity) {
        if (fetchActivityDelegate(activity) != null) {
            fetchActivityDelegate(activity)!!.onStop()
        }
    }

    override fun onActivitySaveInstanceState(activity: Activity, outState: Bundle) {
        if (fetchActivityDelegate(activity) != null) {
            fetchActivityDelegate(activity)!!.onSaveInstanceState(activity, outState)
        }
    }

    override fun onActivityDestroyed(activity: Activity) {
        activity?.let { removeActivity(it) }
        if (fetchActivityDelegate(activity) != null) {
            fetchActivityDelegate(activity)!!.onDestroy()
        }
    }

    private fun fetchActivityDelegate(activity: Activity): IActivityDelegate? {
        return if (activity is IActivity) {
            cache[getKey(activity)]
        } else null
    }

    private fun newDelegate(activity: Activity): IActivityDelegate {
        return ActivityDelegateImpl(activity as FragmentActivity)
    }

    private fun getKey(activity: Activity): String {
        return activity.javaClass.name + activity.hashCode()
    }

}